openedx/core/djangoapps/programs/utils.py

```diff
@@ -461,57 +461,77 @@ def _attach_course_run_upgrade_url(self, run_mode):
         """
-        applicable_seat_types = self.data['applicable_seat_types']
+        applicable_seat_types = set(seat for seat in self.data['applicable_seat_types'] if seat != 'credit')
         is_learner_eligible_for_one_click_purchase = self.data['is_program_eligible_for_one_click_purchase']
         skus = []
         bundle_variant = 'full'
+
         if is_learner_eligible_for_one_click_purchase:
-            for course in self.data['courses']:
-                add_course_sku = True
-                course_runs = course.get('course_runs', [])
-                published_course_runs = filter(lambda run: run['status'] == 'published', course_runs)
+            # Reduce the set of required courses to those for which the user doesn't have an active enrollment
+            courses = self._filter_out_courses_with_enrollments(self.data['courses'])
+
+            # Reduce the set of required courses to those for which the user doesn't have an entitlement
+            courses = self._filter_out_courses_with_entitlements(courses)
 
-                if len(published_course_runs) == 1:
-                    for course_run in course_runs:
-                        is_paid_seat = self._check_enrollment_for_user(course_run)
+            if len(courses) < len(self.data['courses']):
+                bundle_variant = 'partial'
 
-                        if is_paid_seat:
-                            add_course_sku = False
-                            break
+            for course in courses:
+                # Open Question: Do we need to consider whether or not the course is published when selling entitlements?
+                entitlements = [entitlement for entitlement in course['entitlements'] if entitlement['mode'].lower() in applicable_seat_types]
+                if entitlements:
+                    # What if there are multiple acceptable entitlement modes for a course? Which one do we use?
+                    skus.append(entitlements[0]['sku'])
+                else:
+                    course_runs = course.get('course_runs', [])
+                    published_course_runs = filter(lambda run: run['status'] == 'published', course_runs)
 
-                    if add_course_sku:
+                    if len(published_course_runs) == 1:
                         for seat in published_course_runs[0]['seats']:
                             if seat['type'] in applicable_seat_types and seat['sku']:
                                 skus.append(seat['sku'])
+                                break
                     else:
-                        bundle_variant = 'partial'
-                else:
-                    # If a course in the program has more than 1 published course run
-                    # learner won't be eligible for a one click purchase.
-                    is_learner_eligible_for_one_click_purchase = False
-                    skus = []
-                    break
+                        # What is this and why is it necessary?
```

Comment:
This `else` allows us to stop iterating through the courses in the program once we find one that breaks one-click purchase eligibility. Setting `is_learner_eligible_for_one_click_purchase` is, however, unnecessary as the following logic sets it to false if `skus` is an empty list.